### PR TITLE
[JIT] remove prim::shape op

### PR DESCRIPTION
**Pull request by eellison**
Desugar prim::shape to aten::size so that passes don't need to reason about both ops. Serialized models still resolve to `prim::shape` so this doesn't break BC.

### CODE REVIEW

**Review by driazati — APPROVED** [review]

**Comment by houseroad** [comment]
Mar 07 00:49:34 test/onnx/test_pytorch_onnx_caffe2.py::TestCaffe2BackendEmbed_opset10::test_tensor_factories_script FAILED [ 85%]

**Comment by eellison** [comment]
@houseroad @lara-hdr @BowenBao does someone mind taking a look at why replacing `prim::shape` with `aten::size` causes this to fail? They are equivalent ops. It's not clear to me if this is the ONNX export or Caffe2 import that is failing, but ideally there would only exist one version of the op instead of both `prim::shape` and `aten::size`.

**Comment by BowenBao** [comment]
After investigation, the error is that the peephole optimization here https://github.com/pytorch/pytorch/blob/903ad90325507bb8856cfa932dd8d43e35b03fbd/torch/csrc/jit/passes/peephole.cpp#L257 is removing `aten::size` in ir graph. This does not happen for `prim::shape`. 
before:
```
graph(%x.1 : Float(2, 3, 4)):
  %1 : int = prim::Constant[value=1]()
  %2 : None = prim::Constant()
  %3 : int = prim::Constant[value=6]() # test_pytorch_onnx_caffe2.py:1446:50
  %4 : int[] = aten::size(%x.1) # <string>:7:9
  %5 : Tensor = aten::zeros(%4, %3, %2, %2, %2) # test_pytorch_onnx_caffe2.py:1446:23
  %6 : int[] = aten::size(%x.1) # <string>:7:9
  %7 : Tensor = aten::ones(%6, %3, %2, %2, %2) # test_pytorch_onnx_caffe2.py:1446:65
  %8 : Tensor = aten::add(%5, %7, %1) # test_pytorch_onnx_caffe2.py:1446:23
  return (%8)
```
after peephole:
```
graph(%x.1 : Float(2, 3, 4)):
  %1 : int = prim::Constant[value=1]()
  %2 : None = prim::Constant()
  %3 : int = prim::Constant[value=6]() # test_pytorch_onnx_caffe2.py:1446:50
  %9 : int[] = prim::Constant[value=[2, 3, 4]]()
  %5 : Tensor = aten::zeros(%9, %3, %2, %2, %2) # test_pytorch_onnx_caffe2.py:1446:23
  %10 : int[] = prim::Constant[value=[2, 3, 4]]()
  %7 : Tensor = aten::ones(%10, %3, %2, %2, %2) # test_pytorch_onnx_caffe2.py:1446:65
  %8 : Tensor = aten::add(%5, %7, %1) # test_pytorch_onnx_caffe2.py:1446:23
  return (%8)
```

Input `%x.1` becomes unused and thus the error. Besides, the pass that replace `aten::size` with constant shape might cause other problems in the future, since in many scenarios we are expecting dynamic input shape at runtime. @houseroad What's your suggestion?

**Comment by eellison** [comment]
@BowenBao thank you for the investigation! 

What peephole optimizations does ONNX actually need ? I think it would be safer in the future to separate those out and only call them. Since mobile is finalizing bytecode, I would like to land this soon. 

How about for now I rename `addmm_fusion_enabled_` to `onnx_export` and guard the optimization?

File: torch/csrc/jit/passes/peephole.cpp [hunk 1]
```diff
@@ -360,7 +360,9 @@ struct PeepholeOptimizeImpl {
           node->output()->replaceAllUsesWith(output);
           changed_ = true;
         }
-      } else if (node->matches("aten::dim(Tensor self) -> int")) {
+        // disabled in onnx export to not constantify shape info
+      } else if (
+          node->matches("aten::dim(Tensor self) -> int") && !onnx_export_) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/peephole.cpp]: For this issue you should update the block for `aten::size(Tensor self) -> int[]`. Yet it appears that `aten::dim` would have the same issue, so this change is good as well.

**Comment by BowenBao** [comment]
@eellison I'm fine with your proposal. Is this flag not used anymore from jit scripting side? We also need to update the pt-onnx symbolic file to support aten::size https://github.com/pytorch/pytorch/blob/82cdd3abae173985c21889522cbe9f7d6dc43dc5/torch/onnx/symbolic_opset9.py#L380

```
def size(g, self, dim=None):
    if dim is None:
        return g.op("Shape", self)
    if sym_help._maybe_get_const(dim, 'i') < 0:
        rank = self.type().dim()
        if rank:
            dim = sym_help._maybe_get_const(dim, 'i') + rank
            dim = g.op("Constant", value_t=torch.tensor(dim))
    return sym_help._size_helper(g, self, dim)
```

File: torch/csrc/jit/passes/peephole.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/peephole.cpp]: That's a separate but related issue, and I don't know if disabling this will break any models and cause this to be unable to land. I would prefer if it were handled in a different PR.

**Comment by eellison** [comment]
> @eellison I'm fine with your proposal. Is this flag not used anymore from jit scripting side? We also need to update the pt-onnx symbolic file to support aten::size
> 
> https://github.com/pytorch/pytorch/blob/82cdd3abae173985c21889522cbe9f7d6dc43dc5/torch/onnx/symbolic_opset9.py#L380
> 
> ```
> def size(g, self, dim=None):
>     if dim is None:
>         return g.op("Shape", self)
>     if sym_help._maybe_get_const(dim, 'i') < 0:
>         rank = self.type().dim()
>         if rank:
>             dim = sym_help._maybe_get_const(dim, 'i') + rank
>             dim = g.op("Constant", value_t=torch.tensor(dim))
>     return sym_help._size_helper(g, self, dim)
> ```

Nope, it's only used for onnx

File: torch/csrc/jit/passes/peephole.cpp [hunk 2]
```diff
@@ -25,7 +25,7 @@ struct PeepholeOptimizeImpl {
       : aliasDb_(nullptr),
         graph_(graph),
         changed_(true),
-        addmm_fusion_enabled_(addmm_fusion_enabled) {
+        onnx_export_(addmm_fusion_enabled) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/peephole.cpp]: If the flag is indeed only used by onnx, please update all `addmm_fusion_enabled` to `onnx_export` to better reflect the semantics.

**Comment by eellison** [comment]
@bowenbao now I'm getting test failures to  `test_tensor_factories_script`. Thoughts on how to proceed ?

**Comment by eellison** [comment]
@BowenBao looks like `test_arange_dynamic` fails as well. I don't know how many other operators require constant propagation of `aten::dim` to export correctly. Maybe we should go the other direction and allow `aten::dim` to constant prop and disable the test that was using `prim::shape` and failed with it on.

**Comment by BowenBao** [comment]
> @BowenBao looks like `test_arange_dynamic` fails as well. I don't know how many other operators require constant propagation of `aten::dim` to export correctly. Maybe we should go the other direction and allow `aten::dim` to constant prop and disable the test that was using `prim::shape` and failed with it on.

The failures are related to `aten::size`. The tests should pass if you skip the peephole pass for `aten::size` for onnx export.

**Comment by eellison** [comment]
Sure, I updated it. I would be surprised if these changes don't break some downstream model somewhere...

**Comment by eellison** [comment]
@BowenBao now there is a seg fault in `test_arange_dynamic`

**Comment by BowenBao** [comment]
> @BowenBao now there is a seg fault in `test_arange_dynamic`

That one looks weird, I'm building locally to check if I can repro.

**Comment by eellison** [comment]
@BowenBao 

I would guess people rarely use x.shape, and mostly use x.size(). shape is a np compatibility shim. i would rather just remove prim::shape then try to change the onnx behavior of x.size() and x.dim() in the same PR. Removing these will almost certainly break something downstream that has depended on constant prop working in order to export.

**Comment by BowenBao** [comment]
@eellison could you create a separate PR that includes the fixes discussed for disabled tests? We should merge these together. @houseroad @spandantiwari

**Comment by eellison** [comment]
It looks like someone else in onnx is trying to fold prim::shape, https://github.com/pytorch/pytorch/pull/35386  , which this PR also accomplishes.

**Comment by BowenBao** [comment]
> It looks like someone else in onnx is trying to fold prim::shape, #35386 , which this PR also accomplishes.

That PR is only folding on constants. BTW you mentioned in the skip test that some other tests relies on folding enabled, could you point to these exact tests?

**Comment by eellison** [comment]
@BowenBao I have created a separate PR to disable size optimizations, waiting to see if everythnig works well.
`test_arange_dynamic` and `test_tensor_factories_script` are the two tests that would break if you rewrite the calls to `x.shape` to `x.size()`.

**Comment by eellison** [comment]
@BowenBao when I disable test failures in combination with this PR i get seg faults: https://github.com/pytorch/pytorch/pull/35401

**Comment by spandantiwari** [comment]
> @BowenBao when I disable test failures in combination with this PR i get seg faults: #35401

@eellison - I am wondering why we are disabling the tests. Would it possible to fix the test failures in this PR? 

cc: @houseroad

**Comment by eellison** [comment]
Basically the situation is that
- `x.shape` was used for tests that relied on not constant folding
- `x.size()` was used for tests that relied on constant folding

`size` and `shape` are equivalent ops. One was added as a numpy compatibliity shim, but as far as the IR goes, we should not be representing them any differently. If we unify the behavior between the two ops then tests will break no matter what. 

Scripted `x.shape` calls are pretty infrequent; it's never called within the pytorch standard library. To fix the tests that this PR breaks we would have to disable constant folding, however this also causes test failures: https://github.com/pytorch/pytorch/pull/35401. I also think disabling constant folding will have a much large downstream affect for models that rely on it in order to clean up code for exporting.
